### PR TITLE
bpo-29881: Add new C API for static variables

### PR DESCRIPTION
**Pull request by vstinner**
* New type: _Py_StaticVar
* New macro _Py_STATICVAR(var) to declare a variable
* New macro _PY_STATICVAR_INIT(var, expr) to initialize a variable
  once
* New function _PyStaticVar_Set() to explicitly set a variable once
  to initialize it
* New _PyStaticVar_Fini() function clearing all references at exit

### CODE REVIEW

File: Modules/arraymodule.c [hunk 1]
```diff
@@ -2124,20 +2124,19 @@ array_array___reduce_ex__(arrayobject *self, PyObject *value)
...
     int typecode = self->ob_descr->typecode;
     int mformat_code;
-    static PyObject *array_reconstructor = NULL;
+    _Py_STATICVAR(array_reconstructor);
     long protocol;
     _Py_IDENTIFIER(_array_reconstructor);
     _Py_IDENTIFIER(__dict__);
 
-    if (array_reconstructor == NULL) {
+    if (array_reconstructor.obj == NULL) {
```
Comment (serhiy-storchaka) [inline Modules/arraymodule.c:2132]: `_PY_STATICVAR_INIT()` can be used if extract the code for getting `array._array_reconstructor` into a separate function.

File: Python/compile.c [hunk 2]
```diff
@@ -213,7 +213,7 @@ static int compiler_async_comprehension_generator(
                                       expr_ty elt, expr_ty val, int type);
 
 static PyCodeObject *assemble(struct compiler *, int addNone);
-static PyObject *__doc__;
+_Py_STATICVAR(__doc__);
```
Comment (serhiy-storchaka) [inline Python/compile.c:216]: `_Py_IDENTIFIER()` can be used.

File: Python/compile.c [hunk 3]
```diff
@@ -1452,14 +1450,14 @@ compiler_mod(struct compiler *c, mod_ty mod)
 {
     PyCodeObject *co;
     int addNone = 1;
-    static PyObject *module;
-    if (!module) {
-        module = PyUnicode_InternFromString("<module>");
-        if (!module)
-            return NULL;
+    _Py_STATICVAR(module);
```
Comment (serhiy-storchaka) [inline Python/compile.c:1453]: `_Py_static_string()` can be used.

File: Python/compile.c [hunk 4]
```diff
@@ -2639,12 +2637,10 @@ compiler_from_import(struct compiler *c, stmt_ty s)
 
     PyObject *names = PyTuple_New(n);
     PyObject *level;
-    static PyObject *empty_string;
+    _Py_STATICVAR(empty_string);
```
Comment (serhiy-storchaka) [inline Python/compile.c:2640]: `Py_static_string()` can be used.

File: Python/sysmodule.c [hunk 5]
```diff
@@ -162,7 +162,7 @@ sys_displayhook(PyObject *self, PyObject *o)
     PyInterpreterState *interp = PyThreadState_GET()->interp;
     PyObject *modules = interp->modules;
     PyObject *builtins;
-    static PyObject *newline = NULL;
+    _Py_STATICVAR(newline);
```
Comment (serhiy-storchaka) [inline Python/sysmodule.c:165]: `Py_static_string()` can be used.

File: Python/compile.c [hunk 6]
```diff
@@ -2709,17 +2705,18 @@ compiler_from_import(struct compiler *c, stmt_ty s)
 static int
 compiler_assert(struct compiler *c, stmt_ty s)
 {
-    static PyObject *assertion_error = NULL;
+    _Py_STATICVAR(assertion_error);
```
Comment (serhiy-storchaka) [inline Python/compile.c:2708]: `_Py_IDENTIFIER()` can be used.

File: Python/import.c [hunk 7]
```diff
@@ -1721,33 +1721,33 @@ PyImport_ReloadModule(PyObject *m)
 PyObject *
 PyImport_Import(PyObject *module_name)
 {
-    static PyObject *silly_list = NULL;
-    static PyObject *builtins_str = NULL;
-    static PyObject *import_str = NULL;
+    _Py_STATICVAR(empty_list);
+    _Py_STATICVAR(builtins_str);
```
Comment (serhiy-storchaka) [inline Python/import.c:1725]: `_Py_IDENTIFIER()` can be used.

File: Python/import.c [hunk 8]
```diff
@@ -1721,33 +1721,33 @@ PyImport_ReloadModule(PyObject *m)
...
+    _Py_STATICVAR(import_str);
```
Comment (serhiy-storchaka) [inline Python/import.c:1726]: `_Py_IDENTIFIER()` can be used.

File: Modules/arraymodule.c (hunk 1 above)
Comment (vstinner) [reply Modules/arraymodule.c:2132]: done in PR #780.

File: Python/compile.c (hunk 2 above)
Comment (vstinner) [reply Python/compile.c:216]: Right, I know, it was more to show how the API can be used.

File: Python/compile.c (hunk 3 above)
Comment (vstinner) [reply Python/compile.c:1453]: ditto

**Comment by vstinner** [comment]
Abandonned in favor PR #780 which has a simpler API.

**Pull request closed without merging**